server.go

```diff
@@ -21,6 +46,70 @@ type Server struct {
 	PublicKeyHandler    PublicKeyHandler    // public key authentication handler
 	PtyCallback         PtyCallback         // callback for allowing PTY sessions, allows all if nil
 	PermissionsCallback PermissionsCallback // optional callback for setting up permissions
+
+	// Internal fields. Note that the zero value for these should be a state we
+	// can detect so the Server can still be instantiated using &Server{}.
+
+	listener net.Listener    // listener currently being used by Serve, nil otherwise
+	mu       sync.Mutex      // general lock around the Server state
+	wg       *sync.WaitGroup // WaitGroup to track the number of connections still running
+	state    serverState     // tracks if the server is running or not
+	doneChan chan struct{}   // tracks if we want to exit
+	killChan chan struct{}   // tracks if we're blindly killing connections
+}
+
+// getKillChan is used to get or create the kill channel
+func (srv *Server) getKillChan() chan struct{} {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+
+	if srv.killChan == nil {
+		srv.killChan = make(chan struct{})
+	}
+
+	return srv.killChan
+}
+
+// lockedCloseKillChan will close the kill channel if it hasn't already been
+// closed. Note that this assumes srv.mu is held.
+func (srv *Server) lockedCloseKillChan() {
+	if srv.killChan != nil {
+		close(srv.killChan)
+		srv.killChan = nil
+	}
+}
+
+// getKillChan is used to get or create the done channel
+func (srv *Server) getDoneChan() chan struct{} {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+
+	if srv.doneChan == nil {
+		srv.doneChan = make(chan struct{})
+	}
+
+	return srv.doneChan
+}
+
+// lockedCloseDoneChan will close the done channel if it hasn't already been
+// closed. Note that this assumes srv.mu is held.
+func (srv *Server) lockedCloseDoneChan() {
+	if srv.doneChan != nil {
+		close(srv.doneChan)
+		srv.doneChan = nil
+	}
+}
+
+// getKillChan is used to get or create the wait group
+func (srv *Server) getWaitGroup() *sync.WaitGroup {
```

Comment:
I've removed this helper function, but I still think it should be a pointer, otherwise there's a race condition between shutting down a server and starting a new one with the same Server object.